text/0135-compressed-blob-prefixes.md

```diff
@@ -0,0 +1,87 @@
+
+This proposal is intended to support future work enabling Polkadot to execute PolkaVM and, more generally, other-than-Wasm parachain runtimes, and allow developers to introduce arbitrary compression methods seamlessly in the future.
+
+## Stakeholders
+
+Node developers are the main stakeholders for this proposal. It also creates a foundation on which parachain runtime developers will build.
+
+## Explanation
+
+### Overview
+
+The current approach to compressing binary blobs involves using `zstd` compression, and the resulting compressed blob is prefixed with a unique 64-bit magic value specified in that subsection. The same procedure is used to compress both Wasm code blobs and proofs-of-validity. Currently, having solely a compressed blob, it's impossible to tell what's inside it without decompression, a Wasm blob, or a PoV. That doesn't cause problems in the current protocol, as Wasm blobs and PoV blobs take completely different execution paths in the code.
+
+The changes proposed below are intended to define the means for distinguishing compressed blob types in a backward-compatible and future-proof way.
+
+It is proposed to introduce an open list of 64-bit prefixes, each representing a compressed blob of a specific type compressed with a specific compression method. The currently used prefix becomes deprecated and will be removed or reused when it is no longer in use.
+
+The proposed list of prefixes to support the current as well as currently known future work follows:
+
+| Prefix name | Prefix bytes | Description |
+| -- | -- | -- |
+| `CBLOB_ZSTD_LEGACY`      | 82, 188, 83, 118, 70, 219, 142, 5 | Wasm code blob or PoV, zstd-compressed |
+| `CBLOB_ZSTD_POV`         | 82, 188, 83, 118, 70, 219, 142, 6 | Proof-of-validity, zstd-compressed     |
+| `CBLOB_ZSTD_WASM_CODE`   | 82, 188, 83, 118, 70, 219, 142, 7 | Wasm code blob, zstd-compressed        |
+| `CBLOB_ZSTD_PVM_CODE`    | 82, 188, 83, 118, 70, 219, 142, 8 | PolkaVM code blob, zstd-compressed     |
+
+No runtime code changes should be needed to imnplement this proposal. Node-side changes are trivial; a PoC already implemented as a part of [SDK PR#6704](https://github.com/paritytech/polkadot-sdk/pull/6704) may be used as an example.
+
+### Timeline
+
+1. The proposed prefix changes are implemented and released. No logic changes yet;
+2. After the supermajority of production networks' nodes upgrades, one more change is released that adds `CBLOB_ZSTD_WASM_CODE` prefix instead of `CBLOB_ZSTD_LEGACY` when compiling and compressing Wasm parachain runtimes, and `CBLOB_ZSTD_POV` instead of `CBLOB_ZSTD_LEGACY` when compressing PoVs;
+3. Conservatively, wait until no more PVFs prefixed with `CBLOB_ZSTD_LEGACY` remain on-chain. That may take quite some time. Alternatively, create a migration that alters prefixes of existing blobs;
+4. Removing `CBLOB_ZSTD_LEGACY` prefix will be possible after all the collators in all the networks switch to compression with `CBLOB_ZSTD_POV` prefix.
```

Comment:
*Every* parachain node also needs to upgrade to support the new prefixes. We will need to think on how to nudge the people to switch to this option, but it will take some time. Which makes the coordination more complicated.

Reply:
Then I have to ask again, why don't you want to avoid this problem altogether by reusing the existing prefix for PoVs?

Reply:
It is not about PoVs. When a runtime would use a new prefix, old nodes would fail to decompress and thus stall. The Substrate executor is also using this prefix to detect the compressed runtime.

Reply:
Yeah now I got it at last. You're right, there's no point in reusing the legacy prefix in that case. We should incentivize people to upgrade, and the best incentive may be delivering PolkaVM runtimes and lowering PVF deposits for them.

Reply:
Are there any downsides on having `CBLOB_ZSTD_LEGACY` stay for a while? All the nodes will be forced to upgrade to JAM when it is live and can we do it at that time or that's too long?

Reply:
No _real_ downsides indeed, just a reluctance to leave any legacy behind if that can be avoided. I come from the Perl world, and legacy is my pain point :sweat_smile: